### DISCUSSION TITLE
Named return values or exit with local variable?

### DISCUSSION THREAD

**Discussion by PeterConijn** [microsoft/alguidelines#247] [BC Patterns]
This is something that has always irked me a bit. 

Since version (look up version, red.) we have the ability to have named return values. I have, however, noticed that Microsoft seems to avoid using them, opting instead for exiting a local variable. 

There is no pattern* for this at the moment, but which one is the advised preference?

*) This is not about using a named variable vs. exiting with a result directly (e.g. exit(not Customer.IsEmpty()). This is specifically when a variable _is_ required: do we prefer a local variable or a named return value?

Named variable
```AL
local procedure GetCircumference(Radius: Decimal) Circumference: Decimal
var
    Math: Codeunit Math;
begin
    if Radius <= 0 then
        exit(0);

    Circumference := Math.Pi() * 2 * Radius;

    OnAfterGetCircumference(Circumference);
end;
```

Return local variable:
```AL
local procedure GetCircumference(Radius: Decimal): Decimal
var
    Math: Codeunit Math;
    Circumference: Decimal;
begin
    if Radius <= 0 then
        exit(0);

    Circumference := Math.Pi() * 2 * Radius;

    OnAfterGetCircumference(Circumference);
    exit(Circumference);
end;
```

Note: I seem to recall @ajkauffmann once stating that one of the two is more memory-efficient. Perhaps that can be taken into account as well.

**Comment by NKarolak** [comment]
I clearly prefer a named returned variable:
1. No doubt what will be returned.
2. Shorter code because an exit statement is not needed.
3. A must if filtered records are to be returned (see post by [waldo](https://www.waldo.be/2021/07/06/the-complexity-of-complex-return-types/), last section before Conclusion). 
4. As the named variable saves an additional internal assignment, I think it is also performance-wise the better solution.
